Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs

```diff
@@ -0,0 +1,606 @@
+            if (_previousTileIteration == _activeExplosion.CurrentIteration)
+                return;
+
+            _previousTileIteration = _activeExplosion.CurrentIteration;
+            RaiseNetworkEvent(new ExplosionOverlayUpdateEvent(_explosionCounter, _previousTileIteration + 1));
+            return;
+        }
+
+        if (_explosionQueue.Count > 0)
+            return;
+
+        // We have finished processing all explosions. Clear client explosion overlays
+        RaiseNetworkEvent(new ExplosionOverlayUpdateEvent(_explosionCounter, int.MaxValue));
+
+        //wakey wakey
+        _nodeGroupSystem.Snoozing = false;
+    }
+
+    /// <summary>
+    ///     Determines whether an entity is blocking a tile or not. (whether it can prevent the tile from being uprooted
+    ///     by an explosion).
+    /// </summary>
+    /// <remarks>
+    ///     Used for a variation of <see cref="TurfHelpers.IsBlockedTurf()"/> that makes use of the fact that we have
+    ///     already done an entity lookup on a tile, and don't need to do so again.
+    /// </remarks>
+    public bool IsBlockingTurf(EntityUid uid)
+    {
+        if (EntityManager.IsQueuedForDeletion(uid))
+            return false;
+
+        if (!TryComp(uid, out IPhysBody? body))
+            return false;
+
+        return body.CanCollide && body.Hard && (body.CollisionLayer & (int) CollisionGroup.Impassable) != 0;
+    }
+
+    /// <summary>
+    ///     Find entities on a grid tile using the EntityLookupComponent and apply explosion effects. 
+    /// </summary>
+    /// <returns>True if the underlying tile can be uprooted, false if the tile is blocked by a dense entity</returns>
+    internal bool ExplodeTile(EntityLookupComponent lookup,
+        IMapGrid grid,
+        Vector2i tile,
+        float throwForce,
+        DamageSpecifier damage,
+        MapCoordinates epicenter,
+        HashSet<EntityUid> processed,
+        string id)
+    {
+        var gridBox = new Box2(tile * grid.TileSize, (tile + 1) * grid.TileSize);
+
+        // get the entities on a tile. Note that we cannot process them directly, or we get
+        // enumerator-changed-while-enumerating errors.
+        List<(EntityUid, TransformComponent?) > list = new();
+        var xformQuery = EntityManager.GetEntityQuery<TransformComponent>();
+
+        EntityUidQueryCallback callback = uid =>
+        {
+            if (processed.Contains(uid))
+                return;
+
+            var xform = xformQuery.GetComponent(uid);
+
+            if (xform.ParentUid != grid.GridEntityId)
+            {
+                // Not parented to grid. Likely in a container.
+                if (_containerSystem.IsEntityInContainer(uid, xform))
+                    return;
+            }
+
+            list.Add((uid, xform));
+        };
+
+        _entityLookup.FastEntitiesIntersecting(lookup, ref gridBox, callback);
```

Comment:
Changed to a TryGet transform, because that at least checks if the transform comp has been deleted.

> the lambda still allocates so it's not heaps better than an IEnumerable.

What would be the best way of this? AFAIK you still need to store entities in a list/set at some point before dealing damage, otherwise you get modified-while-enumerating errors.

and yeah the container & lookup changes will be good.